torch/lib/THD/base/channels/command_channel/CommandChannel.cpp

```diff
@@ -0,0 +1,137 @@
+#include "CommandChannel.hpp"
+#include "../ChannelEnvVars.hpp"
+#include "../ChannelUtils.hpp"
+
+#include <unistd.h>
+#include <climits>
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <memory>
+#include <string>
+#include <stdexcept>
+#include <utility>
+
+namespace thd {
+namespace {
+
+void sendMessage(int socket, std::unique_ptr<rpc::RPCMessage> msg) {
+  auto& bytes = msg.get()->bytes();
+  std::uint64_t msg_length = static_cast<std::uint64_t>(bytes.length());
+
+  send_bytes<std::uint64_t>(socket, &msg_length, 1);
+  send_bytes<std::uint8_t>(
+    socket,
+    reinterpret_cast<const std::uint8_t*>(bytes.data()),
+    msg_length
+  );
+}
+
+std::unique_ptr<rpc::RPCMessage> receiveMessage(int socket) {
+  std::uint64_t msg_length;
+  recv_bytes<std::uint64_t>(socket, &msg_length, 1);
+
+  std::unique_ptr<std::uint8_t[]> bytes(new std::uint8_t[msg_length]);
+  recv_bytes<std::uint8_t>(socket, bytes.get(), msg_length);
+
+  return std::unique_ptr<rpc::RPCMessage>(
+    new rpc::RPCMessage(reinterpret_cast<char*>(bytes.get()), msg_length)
+  );
+}
+
+} // anonymous namespace
+
+MasterCommandChannel::MasterCommandChannel()
+  : _rank(0)
+{
+  std::uint32_t world_size;
+  std::tie(_port, world_size) = load_master_env();
+
+  _sockets.resize(world_size);
+  std::fill(_sockets.begin(), _sockets.end(), -1);
+}
+
+MasterCommandChannel::~MasterCommandChannel() {
+  for (auto socket : _sockets) {
+    if (socket != -1)
+      ::close(socket);
+  }
+}
+
+bool MasterCommandChannel::init() {
+  // listen on workers
+  std::tie(_sockets[0], std::ignore) = listen(_port);
+
+  int socket;
+  std::uint32_t rank;
+  for (std::size_t i = 1; i < _sockets.size(); ++i) {
+    std::tie(socket, std::ignore) = accept(_sockets[0]);
+    recv_bytes<std::uint32_t>(socket, &rank, 1);
+    _sockets.at(rank) = socket;
+  }
+
+  /* Sending confirm byte is necessary to block workers until all remaing workers
+   * connect. This necessesity comes from case when worker finishes connecting
+   * to command channel and start connecting to data channel. Since master
```

Comment:
I decided to keep that but change description. I think that even this can cause some small performance downgrade, it is a good idea to 'test' the connection and barrier workers. It also won't cause any problems in future if we decide to change order of channels initialization.